egs/wsj/s5/utils/scoring/prepare_calibration_data.py

```diff
@@ -0,0 +1,151 @@
+except IndexError:
+  # Missing 'score_tag' column, add 'U' everywhere,
+  ctm = np.loadtxt(ctm_file,dtype='object,object,f8,f8,object,f8')
+  ctm = [(f, chan, beg, dur, wrd, conf, 'U') for (f, chan, beg, dur, wrd, conf) in ctm]
+  ctm = np.array(ctm,dtype='object,object,f8,f8,object,f8,object')
+
+# Build the targets,
+if o.conf_targets != '':
+  targets = []
+  for (f, chan, beg, dur, wrd, conf, score_tag) in ctm:
+    # Skip words we don't know if being correct, 
+    if score_tag == 'U': continue 
+    # Build the key,
+    key = "%s^%s^%.2f^%.2f^%s,%.2f,%s" % (f, chan, beg, dur, wrd, conf, score_tag)
+    # Build the target,
+    tgt = 1 if score_tag == 'C' else 0 # Correct = 1, else 0,
+    # Add to list,
+    targets.append((key,tgt))
+  # Store the targets,
+  np.savetxt(o.conf_targets, np.array(targets,dtype='object,i'), fmt='%s %d')
+
+# Segments dicionart is indexed by utterance,
+if o.segments != '':
+  segments = { utt:(reco,beg,end) for (utt,reco,beg,end) in np.loadtxt(o.segments,dtype='object,object,f8,f8') }
+  # Optionally apply 'reco2file_and_channel' mapping,
+  if o.reco2file_and_channel != '':
+    reco2file_and_channel = { reco:(file,chan) for (reco,file,chan) in np.loadtxt(o.reco2file_and_channel,dtype='object,object,object') }
+  else:
+    # Or trivial self-mapping using 2nd column in 'segments', (default channel is '1'),
+    reco2file_and_channel = { reco:(reco,'1') for reco in np.unique(np.array(segments.values())['f0']) } 
+
+### Load the per-frame lattice-depth,
+depths = dict()
+if o.segments == '':
+  # In simpler case, the 1st column of 'CTM' are the utterance keys,
+  # which simplifies the lookup of the lattice-dephts.
+  with open(depths_file,'r') as f:
+    for l in f:
+      utt,d = l.split(' ',1)
+      depths[(utt,'1')] = np.array(d.split(),dtype='int')
+else:
+  # In difficult case, the 'CTM' was mapped by 'utils/convert_ctm.pl'.
+  # The depth keys are the utterance-id's, so we fit the depths into 'long' 
+  # vectors covering the complete unsegmented audio channel.
+  #
+  # Getting lengths of unsegmented audio channels (i.e. the end of last segment),
+  reco_beg_end_sort = np.sort(np.array(segments.values(),dtype='object,f8,f8'), order=['f0','f2'])
+  reco_beg_end_lastseg = reco_beg_end_sort[np.append(reco_beg_end_sort['f0'][1:] != reco_beg_end_sort['f0'][:-1],[True])]
+  # Create buffers for the depths,
+  for (reco,beg,end) in reco_beg_end_lastseg:
+    frame_total = 1 + int(np.rint(100*end))
+    depths[reco2file_and_channel[reco]] = np.zeros(frame_total, dtype='int')
+  # Load the depths (ASCII), fill the buffer with the depths,
+  with open(depths_file,'r') as f:
+    for l in f:
+      utt,d = l.split(' ',1)
+      d = np.array(d.split(),dtype='int')
+      frame_begin = int(np.rint(100*segments[utt][1]))
+      depths[reco2file_and_channel[segments[utt][0]]][frame_begin:frame_begin+len(d)] = d
+
+# Load the unigram probabilities in 10log from ARPA,
+import gzip, re
```

Comment:
In order to make it more general, could we let the input to this script just be a file containing the unigram probabilities in some very simple format, and put on the caller the responsibility to get this from the arpa?  That way it will be more adaptable to non-arpa-format language models.